graphix/optimization.py

```diff
@@ -362,6 +371,186 @@ def ensure_neighborhood(node: Node) -> None:
+        cf = CausalFlow(og, dict(correction_function), partial_order_layers)
+        cf.check_well_formed()  # Raises a `FlowError` if the partial order and the correction function are not compatible, or if a measured node is corrected by more than one node.
+        return cf
+
+    def extract_gflow(self) -> GFlow[Measurement]:
+        """Extract the generalized flow (gflow) structure from the current measurement pattern.
+
+        This method does not call the flow-extraction routine on the underlying open graph, but constructs the gflow from the pattern corrections instead.
+
+        Returns
+        -------
+        flow.GFlow[Measurement]
+            The gflow associated with the current pattern.
+
+        Raises
+        ------
+        FlowError
+            If the pattern is empty or if the extracted structure does not satisfy
+            the well-formedness conditions required for a valid gflow.
+        ValueError
+            If `N` commands in the pattern do not represent a |+⟩ state or if the pattern corrections form closed loops.
+
+        Notes
+        -----
+        This method makes use of :func:`StandardizedPattern.extract_partial_order_layers` which computes the pattern's direct acyclical graph (DAG) induced by the corrections and returns a particular layer stratification (obtained by doing a topological sort on the DAG). Further, it constructs the pattern's induced correction function from :math:`M` and :math:`X` commands.
+        In general, there may exist various layerings which represent the corrections of the pattern. To ensure that a given layering is compatible with the pattern's induced correction function, the partial order must be extracted from a standardized pattern. Commutation of entanglement commands with X and Z corrections in the standardization procedure may generate new corrections, which guarantees that all the topological information of the underlying graph is encoded in the extracted partial order.
+        """
+        correction_function: dict[int, set[int]] = defaultdict(set)
+        pre_measured_nodes = set(self.results.keys())  # Not included in the flow.
+
+        for m in self.m_list:
+            if m.plane in {Plane.XZ, Plane.YZ}:
+                correction_function[m.node].add(m.node)
+            correction_function = _update_corrections(m.node, m.s_domain - pre_measured_nodes, correction_function)
+
+        for node, domain in self.x_dict.items():
+            correction_function = _update_corrections(node, domain - pre_measured_nodes, correction_function)
+
+        og = (
+            self.extract_opengraph()
+        )  # Raises a `ValueError` if `N` commands in the pattern do not represent a |+⟩ state.
+        partial_order_layers = (
+            self.extract_partial_order_layers()
+        )  # Raises a `ValueError` if the pattern corrections form closed loops.
+        gf = GFlow(og, dict(correction_function), partial_order_layers)
```

Comment:
`correction_function` is already a `dict`.
```suggestion
        gf = GFlow(og, correction_function, partial_order_layers)
```